[Writing Tools] Writing Tools context menu item may still show up even when using UIWritingToolsBehaviorNone

rr-codes opened the pull request:
#### 63d8246828b2162cb11803da40cb479eca9922b1
<pre>
[Writing Tools] Writing Tools context menu item may still show up even when using UIWritingToolsBehaviorNone
<a href="https://bugs.webkit.org/show_bug.cgi?id=276040">https://bugs.webkit.org/show_bug.cgi?id=276040</a>
<a href="https://rdar.apple.com/130825002">rdar://130825002</a>

Reviewed by Wenson Hsieh.

Implement the `writingToolsBehavior` property of the `{UI|NS}TextInputTraits` protocol on WKContentViewInteraction and WKWebView,
rather than set it on the `writingToolsBehavior` property on the UITextInputTraits object itself on iOS.

This also has the advantage of unifying the behavior across the platforms.

* Source/WebCore/PAL/pal/spi/ios/UIKitSPI.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView writingToolsBehavior]):
(-[WKWebView wantsWritingToolsInlineEditing]): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/API/mac/WKWebViewMac.mm:
(-[WKWebView _web_wantsWritingToolsInlineEditing]): Deleted.
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView canPerformActionForWebView:withSender:]):
(-[WKContentView _updateTextInputTraits:]):
(-[WKContentView writingToolsBehavior]):
(-[WKContentView _startWritingToolsForWebView:]): Deleted.
(-[WKContentView wantsWritingToolsInlineEditing]): Deleted.
* Source/WebKit/UIProcess/ios/WKExtendedTextInputTraits.h:
* Source/WebKit/UIProcess/ios/WKExtendedTextInputTraits.mm:
(-[WKExtendedTextInputTraits restoreDefaultValues]):
(-[WKExtendedTextInputTraits restoreDefaultWritingToolsBehaviorValue]): Deleted.
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::selectionDidChange):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
(-[WritingToolsWKWebView writingToolsBehaviorForTesting]):
(TEST(WritingTools, WantsInlineEditing)):
(TEST(WritingTools, WritingToolsBehaviorNonEditableWithSelection)):
(TEST(WritingTools, WritingToolsBehaviorWithNoSelection)):
(TEST(WritingTools, WritingToolsBehaviorEditableWithSelection)):
(TEST(WritingTools, APIWithBehaviorNone)):
(TEST(WritingTools, APIWithBehaviorDefault)):
(TEST(WritingTools, APIWithBehaviorComplete)):
(TEST(WritingTools, ShowAffordance)):
(TEST(WritingTools, ShowAffordanceForMultipleLines)):

Canonical link: <a href="https://commits.webkit.org/280518@main">https://commits.webkit.org/280518@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4adc6ef635eeeaac96fcccaa10ec398df3e2e01a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/56867 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36195 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9341 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/60486 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/7310 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/58995 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/43818 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/7500 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46059 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5126 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/58897 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/33997 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49077 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/26917 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/30777 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/6408 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6314 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/52733 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/6679 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/62168 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/780 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/6783 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/53316 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/783 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49132 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/53351 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/659 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8464 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32024 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33109 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/34194 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/32855 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
```diff
@@ -179,6 +180,15 @@ - (PlatformWritingToolsAllowedInputOptions)writingToolsAllowedInputOptionsForTes
+- (PlatformWritingToolsBehavior)writingToolsBehaviorForTesting
+{
+#if PLATFORM(IOS_FAMILY)
+    return [(id<UITextInput>)[self textInputContentView] writingToolsBehavior];
```
I think the cast here is unnecessary.

rr-codes replied:
I did it for consistency with the implementation of `writingToolsAllowedInputOptionsForTesting`. I'll change it

rr-codes committed:
[Writing Tools] Writing Tools context menu item may still show up even when using UIWritingToolsBehaviorNone
https://bugs.webkit.org/show_bug.cgi?id=276040
rdar://130825002

Reviewed by Wenson Hsieh.

Implement the `writingToolsBehavior` property of the `{UI|NS}TextInputTraits` protocol on WKContentViewInteraction and WKWebView,
rather than set it on the `writingToolsBehavior` property on the UITextInputTraits object itself on iOS.

This also has the advantage of unifying the behavior across the platforms.

* Source/WebCore/PAL/pal/spi/ios/UIKitSPI.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView writingToolsBehavior]):
(-[WKWebView wantsWritingToolsInlineEditing]): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/API/mac/WKWebViewMac.mm:
(-[WKWebView _web_wantsWritingToolsInlineEditing]): Deleted.
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView canPerformActionForWebView:withSender:]):
(-[WKContentView _updateTextInputTraits:]):
(-[WKContentView writingToolsBehavior]):
(-[WKContentView _startWritingToolsForWebView:]): Deleted.
(-[WKContentView wantsWritingToolsInlineEditing]): Deleted.
* Source/WebKit/UIProcess/ios/WKExtendedTextInputTraits.h:
* Source/WebKit/UIProcess/ios/WKExtendedTextInputTraits.mm:
(-[WKExtendedTextInputTraits restoreDefaultValues]):
(-[WKExtendedTextInputTraits restoreDefaultWritingToolsBehaviorValue]): Deleted.
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::selectionDidChange):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
(-[WritingToolsWKWebView writingToolsBehaviorForTesting]):
(TEST(WritingTools, WantsInlineEditing)):
(TEST(WritingTools, WritingToolsBehaviorNonEditableWithSelection)):
(TEST(WritingTools, WritingToolsBehaviorWithNoSelection)):
(TEST(WritingTools, WritingToolsBehaviorEditableWithSelection)):
(TEST(WritingTools, APIWithBehaviorNone)):
(TEST(WritingTools, APIWithBehaviorDefault)):
(TEST(WritingTools, APIWithBehaviorComplete)):
(TEST(WritingTools, ShowAffordance)):
(TEST(WritingTools, ShowAffordanceForMultipleLines)):

Canonical link: https://commits.webkit.org/280518@main
